### PR TITLE
#16 production ready image

### LINKED ISSUE #16: proposal of example of a production ready Dockerfile and docker-compose.yml
Hi @amato-gianluca ,

Would you like me to build a PR with an example of a Dockerfile and docker-compose.yml ready for production?

Having done the other issue #15, maybe it would be a good idea to provide an example of what could be a production-ready WIMS docker image and its docker-compose.yml.

It would be based on what we would put in production. So other can be benefited and other can provide other issues that they may share with this repo.

Let me know.

Thanks a lot for your work,

Jordi

### PR DESCRIPTION
**Pull request by jpahullo**
closes #16 

@amato-gianluca ,

It is a first version for a production-ready version of the WIMS image.

I tried several options, and I think this is the best option and more sustainable version for a production-ready version of the WIMS image. Sincerely, I imagined another way of doing it, but finally I faced the current one.

At least, in our case, we could easily extend it to customize it with our configuration for the production infrastructure.

It is based on the current branch that solves #15, just to be over the latest version of this repository (when #15 is merged). So, do not be afraid with the amount of changes. 

Better focus on the PR related to #15. If you like, you could take a look on the content of the prod/ directory.

Since I made last-minute changes, maybe I broke some things. So, just take a look. Let me double check it locally first. But, your feedback will be very valuable.

Thanks for your time.

Jordi

### CODE REVIEW

**Comment by amato-gianluca** [comment]
There is a conflict in the Makefile. Could you please fix it (I am not particularly expert of make)  ?

**Comment by amato-gianluca** [comment]
This PR includes many files, and some of them are clearly tied to your installation (like `sever.geoloc`). Nonetheless, my intention would be to merge the PR as it is (after resolving the conflict), and later clean it of the configurations which does not seem to be general enough.

Or do you want to suggest a different approach ?

**Comment by jpahullo** [comment]
Hi @amato-gianluca ,

I will check the PR. Since it was based on the other branch, and not master one, it is expected to show conflicts. So, let me check it locally and I will update the PR without conflicts.

Regarding the amount of files: You should actually consider only the content of prod/ directory:

1. Most of them are about the entrypoint and post-install. The revisited 4_wims.conf.sh is thought for a production infrastructure where the number of CPUs can be altered. So, next start will recheck it and update the relevant config.
2. About configuration files and related to my infrastructure: strictly related to our institution is the `geoloc` file you mention. But it had the purpose to quickly adapt it to start it with your own location. I could add a README.md on that directory explaining this, since the `geoloc` file can only contain that. 
3. The security.conf file is a generic configuration for production-ready apache service.
4. Regarding WIMS logs and redirection: The approach is to redirect all logs to journald service, plugin available on docker compose, so that you have a single point with all logs. This is a proposal for a production-ready WIMS instance, thinking on the possibility to have them on several frontends, preventing errors on the process or on its content from writing several frontends to the same file. I redirected all those logs that I identified, but I do think that having the WIMS at full usage, other logs will appear, so I will need to add them to be redirected to the stderr and stdout of the container.
5. The `.connections` directory contains an example of a WIMS integration with other platforms of the UNICE site (not ours). Just to inform that WIMS can be integrated with other platforms, like with Moodle. And you need to set up this kind of files.
6. Finally, the provided `wims.conf` is a template of what you can customize (it is a part, not all WIMS options appear there). Colours were picked from another site (I do not remember which). We do not have yet defined the look&feel for our WIMS. I purposely left empty fields to force edition on its reuse/extension. Other values were set up during my local trials. I do not know if they will be our final settings. Which I set up there also were the commands for the programs to be used by WIMS, like scilab_command and so. Also, I set up some quality settings, since default ones from WIMS produce images with poor quality. They are a good point to start with.

But, above all, it is your repository and your decision. I can simply describe the rationale behind any decision I made, for every set of files I put on it. Please, be sure the proposal for this production-ready WIMS image is simply an option, with a deep thinking on why things are done this way and its consequences on a dockerized environement. In the end, anyone could customize it extending this repo and change whatever they need to. In my opinion, however, it is a good starting point.

In summary:

1. I will update this PR without conflicts.
2. Let me know if you need something else before pushing it again.

Thanks a lot for your time,

Jordi

**Comment by jpahullo** [comment]
Hi @amato-gianluca ,

Simply rebasing to the latest master branch, conflicts were gone.

Let me know if you need something else.

Thanks for your work!

Jordi

**Comment by jpahullo** [comment]
Hi @amato-gianluca  again,

I updated the Makefiles.

I checked that starting from a brand new start of a clone of this repo all works properly. And I detected that .env files, when missing, produced errors and could not start it.

So now, Makefile loads .env-dist always and then loads .env if present, without warnings nor errors.

It will work all ok right now.

Thanks again,

Jordi
